README.md
# LiteRT/SMTP

[![npm version](https://img.shields.io/npm/v/@litert/smtp.svg?colorB=brightgreen)](https://www.npmjs.com/package/@litert/smtp "Stable Version")
[![License](https://img.shields.io/npm/l/@litert/smtp.svg?maxAge=2592000?style=plastic)](https://github.com/litert/smtp/blob/master/LICENSE)
[![GitHub issues](https://img.shields.io/github/issues/litert/smtp.js.svg)](https://github.com/litert/smtp.js/issues)
[![GitHub Releases](https://img.shields.io/github/release/litert/smtp.js.svg)](https://github.com/litert/smtp.js/releases "Stable Release")

The SMTP implement based on LiteRT.

> **Currently only the server-side sending is implemented, which means no delegate works.**

## Installation

```sh
npm install @litert/smtp --save
```

## Requirements

- Node.js v8.x (Or newer)
- TypeScript v3.1.x (Or newer)

## Documents

Preparing yet.

## License

This library is published under [Apache-2.0](./LICENSE) license.


CHANGES.md
# Changes Logs

## v0.2.0

- Added DKIM supports
- Added dynamic domains management.
- Fixed: the BASE64 output should not be more than 78 charactors each line.
- Added tool class HeaderBuilder.

## v0.1.1

- Fixed the arguments of APIs.


src/lib/DKIM.ts
import * as Signs from "@litert/signatures";
import * as C from "./Common";
import * as E from "./Errors";

export interface IDKIMSignOptions {

    headers: string[];

    body: string;

    canonicalization: "simple" | "relaxed" | "simple/simple" | "relaxed/simple" | "simple/relaxed" | "relaxed/relaxed";
}

export interface IDKIMHelper {

    sign(info: IDKIMSignOptions): string;
}

export interface IDKIMOptions {

    pubKey: string;

    privKey: Signs.IAsymmetricKey;

    selector: string;
}

export class DKIMHelper {

    private signer: Signs.ISigner<"base64">;

    public constructor(
        private _options: IDKIMOptions
    ) {

        this.signer = Signs.RSA.createSigner(
            "sha256",
            "",
            this._options.privKey,
            Signs.ERSAPadding.PKCS1_V1_5,
            "base64"
        );
    }

    private _validateCanonicalization(
        canonicalization: IDKIMSignOptions["canonicalization"]
    ): IDKIMSignOptions["canonicalization"] {

        const p = canonicalization.split("/");

        if (p.filter((v) => ["simple", "relaxed"].includes(v)).length !== p.length) {

            throw new E.E_DKIM_INVALID_CANO();
        }

        switch (p.length) {
        case 1: return p + "/simple" as any;
        case 2: return canonicalization;
        default:
            throw new E.E_DKIM_INVALID_CANO();
        }
    }

    public _canonicalizeBody(body: string, algo: "relaxed" | "simple"): string {

        if (algo === "simple") {

            /**
             * simple algo:
             *
             * 1. Remove all empty lines at the end of body.
             * 2. Make sure the body ending with a CRLF.
             */
            return body.replace(/(\r?\n)+$/, "") + C.CRLF;
        }
        else {

            if (!body) {

                return "";
            }

            /**
             * relaxed algo:
             *
             * 1. Replace all tab/space into space.
             * 2. Remove all space/CR/LF at the end of lines.
             * 3. Make sure all lines end with CRLF.
             * 4. Remove all empty lines at the end of body.
             * 5. Make sure the body ending with a CRLF.
             */
            return body.replace(/[ \t]+/g, " ")
            .split("\n")
            .map((l) => l.replace(/[ \r]+$/, ""))
            .join(C.CRLF)
            .replace(/(\r\n)+$/, "") + C.CRLF;
        }
    }

    public _canonicalizeHeaderLine(item: string, algo: "relaxed" | "simple"): string {

        if (algo === "simple") {

            return item;
        }

        let [k, ...v] = item.split(":");

        return k.toLowerCase().trim().replace(/[ \t]+/g, " ") + ":" + v.join(":").replace(
            /\r?\n/g, ""
        ).replace(
            /[ \t]+/g, " "
        ).replace(
            /[ \t]+$/, ""
        ).replace(
            /^[ \t]+/, ""
        );
    }
}


src/lib/Common.ts
/**
 * Copyright 2018 Angus.Fenying
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export interface IPerson {

    /**
     * The display name of the person.
     */
    "name": string;

    /**
     * The mailbox address of the person.
     */
    "address": string;
}

export enum EFormat {

    TEXT,
    HTML
}

export interface IMailOptions {

    /**
     * The sender.
     */
    "from": IPerson;

    /**
     * The recipients.
     */
    "to": IPerson[];

    /**
     * The additional recipients.
     */
    "cc"?: IPerson[];

    /**
     * Reply to the address.
     */
    "replyTo"?: IPerson;

    /**
     * The letter subject.
     */
    "subject": string;

    /**
     * the letter body.
     */
    "body": string;

    /**
     * The format of a letter.
     */
    "format": EFormat;

    /**
     * The unique message id of a letter.
     *
     * __IT WILL AUTO APPEND THE DOMAIN AT THE END.__
     */
    "uuid"?: string;

    /**
     * Set to `false` to disable the DKIM when the domain enabled DKIM.
     *
     * If domain doesn't enable DKIM, this option will be ignored.
     *
     * @default true
     */
    "dkim"?: boolean;
}

export interface IServerSender {

    /**
     * Send a letter.
     *
     * @param info The details of the letter.
     */
    send(info: IMailOptions): Promise<void>;

    /**
     * Close all connections.
     */
    close(): void;

    /**
     * Add a new domain. If the domain already exists, the previous will be overwritten.
     *
     * @param domain The domain info.
     */
    addDomain(domain: string | IDomainOptions): this;

    /**
     * Remove a existing domain.
     *
     * @param domain The domain.
     */
    removeDomain(domain: string): this;

    /**
     * Check if a domain exists.
     *
     * @param domain The domain.
     */
    existDomain(domain: string): boolean;
}

export type TDKIMSigner = (
    headers: string[],
    body: string,
    canonicalization: string,
    selector: string,
    domain: string,
    privateKey: string | Buffer,
    headersToSigned?: string[]
) => string;

export interface IDomainOptions {

    /**
     * The domain. If not lowercase, it will be transform into lowercase.
     */
    domain: string;

    /**
     * The DKIM settings. If omittedc, the DKIM will not be used for this domain.
     */
    dkim?: {

        /**
         * The SELECTOR used for DKIM signature.
         */
        selector: string;

        /**
         * The private RSA key used for DKIM signature.
         */
        privateKey: string | Buffer;

        /**
         * The canonicalization algorithm to be used.
         *
         * @default "relaxed/relaxed"
         */
        canonicalization?: string;

        /**
         * The headers to be signed in DKIM.
         *
         * @default ["from", "to", "cc", "subject", "message-id"]
         */
        headers?: string[];
    };
}

export interface ISenderOptions {

    /**
     * The valid domains for sending letter.
     */
    domains: Array<string | IDomainOptions>;

    /**
     * How long will the DNS query result be cached, in milliseconds.
     *
     * @default 600000 ms
     */
    dnsCache?: number;

    /**
     * The DKIM signer fnction for the sender.
     *
     * If omitted, DKIM will not work.
     */
    dkimSigner?: TDKIMSigner;
}

export const CRLF = "\r\n";

export interface IHeaderBuilder {

    /**
     * Append a piece of text.
     *
     * @param piece The piece of text to be appended.
     */
    append(piece: string): this;

    /**
     * Build the header line.
     */
    toString(): string;
}

export const DEFAULT_CANONICALIZATION = "relaxed/relaxed";

export const DEFAULT_DKIM_SIGN_HEADERS: string[] = [
    "from", "to", "cc", "reply-to",
    "date", "subject", "message-id", "content-type"
];


src/lib/HeaderBuilder.ts
import { CRLF, IHeaderBuilder } from "./Common";

class HeaderBuilder implements IHeaderBuilder {

    private _content: string = "";

    private _indent: string;

    private _lineContentLen: number;

    public constructor(
        name: string,
        private _maxLineLength: number = 80,
        private _indentLength: number = 4
    ) {

        this._content += name + ": ";

        this._indent = " ".repeat(this._indentLength);

        this._lineContentLen = this._maxLineLength - CRLF.length - this._indentLength;
    }

    public append(p: string): this {

        this._content += p;

        return this;
    }

    private _softCut(p: string, maxLen: number): [string, string] {

        let s = p.slice(0, maxLen);

        let c = s.match(/[ \t]/g) as string[];

        if (c) {

            s = s.slice(0, s.lastIndexOf(c[c.length - 1]));
        }
        else {

            c = s.match(/[:;]/g) as string[];

            if (c) {

                s = s.slice(0, s.lastIndexOf(c[c.length - 1]) + 1);
            }
        }

        return [s, p.slice(s.length).trim()];
    }

    public toString(): string {

        const lines: string[] = [];

        let content = this._content;

        if (this._content.length > this._lineContentLen + this._indentLength) {

            let ln: string;

            [ln, content] = this._softCut(content, this._lineContentLen + this._indentLength);

            lines.push(ln);
        }

        while (content.length > this._lineContentLen) {

            let ln: string;

            [ln, content] = this._softCut(content, this._lineContentLen);

            lines.push(this._indent + ln);
        }

        if (content) {

            lines.push(this._indent + content);
        }

        return lines.join(CRLF);
    }
}

export function createHeaderBuilder(

    name: string,
    maxLineLength?: number,
    indentLength?: number
): IHeaderBuilder {

    return new HeaderBuilder(name, maxLineLength, indentLength);
}


src/lib/ServerSender.ts
/**
 * Copyright 2018 Angus.Fenying
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// tslint:disable: no-string-literal

import * as $DNS from "dns";
import * as $Net from "net";
import * as C from "./Common";
import * as E from "./Errors";
import { RawPromise } from "@litert/core";

class Connection {

    private _socket!: $Net.Socket;

    private _connectPromise!: Promise<void>;

    private _buffer!: string;

    private _waiters!: Array<RawPromise<any>>;

    public constructor(
        public readonly mxHost: string,
        public readonly hostname: string,
        private readonly _fromHost: string
    ) {}

    public close(): void {

        if (this._socket) {

            this._socket.end();
        }
    }

    public async connect(): Promise<void> {

        if (this._connectPromise) {

            return this._connectPromise;
        }

        if (this._socket) {

            return;
        }

        await (this._connectPromise = new Promise((resolve, reject) => {

            this._socket = $Net.connect({
                host: this.mxHost,
                port: 25
            })
            .on("connect", () => {

                delete this._connectPromise;

                this._buffer = "";

                this._waiters = [];

                this._socket.removeAllListeners("error");

                this._socket.on("error", (e) => {

                    delete this._socket;
                });

                resolve();

            })
            .on("error", (e) => {

                delete this._socket;
                delete this._connectPromise;
                reject(e);

            })
            .on("data", (data: Buffer) => this._writeBuffer(data.toString()))
            .on("close", () => {

                for (const ret of this._waiters) {

                    ret.reject(new E.E_CONN_LOST());
                }

                this._waiters = [];

                delete this._socket;
            });
        }));

        await this.send(`HELO ${this._fromHost}`);
    }

    private _writeBuffer(chunk: string): void {

        this._buffer += chunk;

        const responses = this._buffer.split(C.CRLF);

        this._buffer = responses.pop() as string;

        let msg: string;

        // tslint:disable-next-line: no-conditional-assignment
        while (msg = responses.shift() as string) {

            const code = parseInt(msg.substr(0, 3));

            let extCode = "";

            if (msg[5] === "." && msg[7] === ".") {

                extCode = msg.slice(4, 9);
                msg = msg.slice(9);
            }
            else {
                msg = msg.slice(4);
            }

            switch (code) {
                case 220: // Server accepted connection.

                    break;

                case 221: // Server closed.

                    this._socket.end();
                    delete this._socket;

                    break;

                case 250: {// Command OK.

                    const ret = this._waiters.shift() as RawPromise<any>;
                    ret.resolve(msg);

                    break;
                }
                case 354: {// Command OK.

                    const ret = this._waiters.shift() as RawPromise<any>;
                    ret.resolve(msg);

                    break;
                }
                case 420: { // Connection timeout.

                    while (1) {

                        const ret = this._waiters.shift() as RawPromise<any>;

                        if (ret) {

                            ret.reject(new E.E_CONN_TIMEOUT());
                        }
                        else {

                            break;
                        }
                    }

                    this._socket.end();
                    delete this._socket;
                    break;
                }
                default: {

                    const ret = this._waiters.shift() as RawPromise<any>;
                    const ERROR = E.ErrorHub.get(code);

                    ret.reject(new ERROR({
                        "metadata": {
                            "details": {
                                "code": extCode,
                                "message": msg
                            }
                        }
                    }));
                }
            }
        }
    }

    public send(command: string): Promise<string> {

        if (!this._socket) {

            throw new E.E_CONN_LOST();
        }

        let ret = new RawPromise<string>();

        this._socket.write(command + C.CRLF);

        this._waiters.push(ret);

        return ret.promise;
    }
}

interface IDNSRecord {

    exchange: string;

    expiringAt: number;
}

class ServerSender implements C.IServerSender {

    private _pool: Record<string, Connection> = {};

    private _dnsCache: Record<string, IDNSRecord> = {};

    private _domains: Record<string, C.IDomainOptions> = {};

    public constructor(
        domains: Array<C.IDomainOptions | string>,
        private _dnsTTL: number = 600000,
        private _dkimSigner?: C.TDKIMSigner
    ) {

        domains.forEach((v) => this.addDomain(v));
    }

    public addDomain(domain: string | C.IDomainOptions): this {

        if (typeof domain === "string") {

            domain = { domain };
        }

        this._domains[domain.domain.toLowerCase()] = {
            "domain": domain.domain.toLowerCase(),
            "dkim": domain.dkim
        };

        return this;
    }

    public removeDomain(domain: string): this {

        delete this._domains[domain.toLowerCase()];

        return this;
    }

    public existDomain(domain: string): boolean {

        return !!this._domains[domain.toLowerCase()];
    }

    public async send(info: C.IMailOptions): Promise<void> {

        const usedDomain = this._extractHostname(
            info.from.address = info.from.address.toLowerCase().trim()
        );

        if (!this._domains[usedDomain]) {

            throw new E.E_USER_NOT_FOUND();
        }

        info.to = info.to.map((v) => ({
            name: v.name,
            address: v.address.toLowerCase().trim()
        }));

        info.cc = (info.cc || []).map((v) => ({
            name: v.name,
            address: v.address.toLowerCase().trim()
        }));

        const recipients: Record<string, C.IPerson[]> = {};

        for (let p of Array.from(new Set([
            ...info.to,
            ...info.cc
        ]))) {

            let hostname = this._extractHostname(p.address);

            if (!recipients[hostname]) {

                recipients[hostname] = [];
            }

            recipients[hostname].push(p);
        }

        // const MAIL_DATA = [
        //     "MIME-Version: 1.0",
        //     `Date: ${new Date().toUTCString()}`,
        //     `From: ${this._wrapRecipient(info.from)}`,
        //     `Delivered-To: ${info.to.map((v) => this._wrapRecipient(v)).join(", ")}`,
        //     `To: ${info.to.map((v) => this._wrapRecipient(v)).join(", ")}`,
        //     ...((): string[] => {

        //         const ret: string[] = [];

        //         if (info.uuid) {

        //             ret.push(`Message-Id: <${info.uuid}@${usedDomain}>`);
        //         }

        //         if (info.cc.length) {

        //             ret.push(`Cc: ${info.cc.map((v) => this._wrapRecipient(v)).join(", ")}`);
        //         }

        //         if (info.replyTo) {

        //             ret.push(`Reply-To: ${this._wrapRecipient(info.replyTo)}`);
        //         }

        //         return ret;
        //     })(),
        //     `Subject: ${this._encodeUTF8(info.subject)}`,
        //     `Content-Type: ${this._getContentType(info.format)}; charset=UTF-8`,
        //     `Content-Transfer-Encoding: base64`,
        //     "",
        //     Buffer.from(info.body).toString("base64"),
        //     "."
        // ].join(C.CRLF);

        const MAIL_DATA: string[] = [];
        const MAIL_HEADERS: Record<string, string> = {};
        const MAIL_BODY = Buffer.from(info.body)
                        .toString("base64")
                        .replace(/(.{78})(?!$)/g, "$1" + C.CRLF);

        MAIL_HEADERS["MIME-Version"] = "1.0";
        MAIL_HEADERS["Date"] = new Date().toUTCString();
        MAIL_HEADERS["From"] = this._wrapPerson(info.from);
        MAIL_HEADERS["To"] = MAIL_HEADERS["Delivered-To"] = info.to.map(
            (v) => this._wrapPerson(v)
        ).join(", ");
        MAIL_HEADERS["From"] = this._wrapPerson(info.from);

        if (info.uuid) {

            MAIL_HEADERS["Message-Id"] = `<${info.uuid}@${usedDomain}>`;
        }

        if (info.cc.length) {

            MAIL_HEADERS["Cc"] = info.cc.map((v) => this._wrapPerson(v)).join(", ");
        }

        if (info.replyTo) {

            MAIL_HEADERS["Reply-To"] = this._wrapPerson(info.replyTo);
        }

        MAIL_HEADERS["Subject"] = this._encodeUTF8(info.subject);
        MAIL_HEADERS["Content-Type"] = `${this._getContentType(info.format)}; charset=UTF-8`;
        MAIL_HEADERS["Content-Transfer-Encoding"] = "base64";

        const domain = this._domains[usedDomain];

        for (const f in MAIL_HEADERS) {

            MAIL_DATA.push(`${f}: ${MAIL_HEADERS[f]}`);
        }

        /**
         * If domain enabled DKIM, use DKIM by default.
         */
        if (domain.dkim && info.dkim !== false) {

            if (!this._dkimSigner) {

                throw new E.E_DKIM_NOT_READY();
            }

            const dkimHeader = this._dkimSigner(
                MAIL_DATA,
                MAIL_BODY,
                domain.dkim.canonicalization || C.DEFAULT_CANONICALIZATION,
                domain.dkim.selector,
                usedDomain,
                domain.dkim.privateKey,
                domain.dkim.headers || C.DEFAULT_DKIM_SIGN_HEADERS
            );

            MAIL_DATA.push(dkimHeader);
        }

        MAIL_DATA.push("");
        MAIL_DATA.push(MAIL_BODY);
        MAIL_DATA.push(".");

        for (const hostname in recipients) {

            let conn = this._pool[hostname];

            if (!conn) {

                conn = new Connection(
                    await this._nsLookUp(hostname),
                    hostname,
                    usedDomain
                );
            }

            await conn.connect();

            await conn.send(`MAIL FROM: <${info.from.address}>`);

            for (let to of recipients[hostname]) {

                await conn.send(`RCPT TO: <${to.address}>`);
            }

            await conn.send(`DATA`);

            await conn.send(MAIL_DATA.join(C.CRLF));
        }
    }

    public close(): void {

        for (let k in this._pool) {

            this._pool[k].close();
        }

        this._pool = {};
    }

    private _extractHostname(addr: string): string {

        return addr.slice(addr.indexOf("@") + 1);
    }

    private _getContentType(type: C.EFormat): string {

        return type === C.EFormat.HTML ? "text/html" : "text/plain";
    }

    private _encodeUTF8(text: string): string {

        return `=?UTF-8?B?${Buffer.from(text).toString("base64")}?=`;
    }

    /**
     * Wrap the recipient/sender info into UTF-8 encoding if necessary.
     *
     * @param p The person info.
     */
    private _wrapPerson(p: C.IPerson) {

        if (/^[-\. \w]+$/.test(p.name)) {

            return `"${p.name}" <${p.address}>`;
        }

        return `${this._encodeUTF8(p.name)} <${p.address}>`;
    }

    /**
     * Find the MX record of determined hostname, with the highest priority.
     *
     * @param hostname The hostname to be looked up.
     */
    private _nsLookUp(hostname: string): Promise<string> {

        let d = this._dnsCache[hostname];

        if (d && d.expiringAt >= Date.now()) {

            return Promise.resolve(d.exchange);
        }

        return new Promise((resolve, reject) => {

            $DNS.resolveMx(hostname, (error, addresses) => {

                if (error) {

                    return reject(new E.E_DNS_ERROR({ metadata: { error } }));
                }

                const theOne = addresses.find(
                    (v) => v.priority === Math.max(...addresses.map((x) => x.priority))
                );

                if (!theOne) {

                    return reject(new E.E_DNS_ERROR());
                }

                if (this._dnsTTL > 0) {

                    this._dnsCache[hostname] = {
                        exchange: theOne.exchange,
                        expiringAt: Date.now() + this._dnsTTL
                    };
                }

                resolve(theOne.exchange);
            });
        });
    }
}

export function createServerSender(opts: C.ISenderOptions): C.IServerSender {

    return new ServerSender(
        opts.domains,
        opts.dnsCache,
        opts.dkimSigner
    );
}
